### PR TITLE
Types class changes for fixed points

### PR DESCRIPTION
**Pull request by randomnetcat**
Formerly part of #3389.
As requested by @axic.

### CODE REVIEW

**Review by axic — DISMISSED** [review]
I think looks OK to merge. The actual fixed point PR can then only focus on implementation details and not about setting the rules.

File: libsolidity/ast/Types.cpp [hunk 1]
```diff
@@ -688,6 +687,11 @@ TypePointer FixedPointType::binaryOperatorResult(Token::Value _operator, TypePoi
 		_other->category() != Category::Integer
 	)
 		return TypePointer();
+
+	if (auto integerType = dynamic_pointer_cast<IntegerType const>(_other)) {
```
Comment (chriseth) [inline libsolidity/ast/Types.cpp]: Style: No braces, or at least on a line of their own.

File: libsolidity/ast/Types.cpp [hunk 2]
```diff
@@ -688,6 +687,11 @@ TypePointer FixedPointType::binaryOperatorResult(Token::Value _operator, TypePoi
...
+		if (integerType->isAddress()) return TypePointer();
```
Comment (chriseth) [inline libsolidity/ast/Types.cpp]: return on a line of its own.

File: libsolidity/ast/Types.cpp [hunk 3]
```diff
@@ -688,6 +687,11 @@ TypePointer FixedPointType::binaryOperatorResult(Token::Value _operator, TypePoi
...
+	}
+
 	auto commonType = Type::commonType(shared_from_this(), _other); //might be fixed point or integer
```
Comment (chriseth) [inline libsolidity/ast/Types.cpp]: I actually think that this should handle the above case already. FixedPointType and Address cannot be converted into each other and thus `commonType` is null.

File: libsolidity/ast/Types.h [hunk 4]
```diff
@@ -396,6 +396,9 @@ class FixedPointType: public Type
 	/// smallest value in general.
 	bigint minIntegerValue() const;
 
+	/// Returns the representation of this type as an integer type. Units are 10 ^ (-fractionalDigits())
```
Comment (chriseth) [inline libsolidity/ast/Types.h]: Perhaps better: return smallest integer type that can hold this type with fractional parts shifted to integers, or something like hat.

File: libsolidity/ast/Types.cpp (hunk 3 above)
Comment (axic) [reply libsolidity/ast/Types.cpp]: `commonType` checks if either side can be converted implicitly to the other:
- `IntegerType` allows implicit conversion to `FixedPoint` if it is not an address
- `RationalNumberType` allows implicit conversion to `FixedPoint`
- `FixedPoint` only allows implicit conversion to other configurations without precision loss of `FixedPoint`

Based on that the entire category check above the `commonType` could be removed here.

**Comment by axic** [comment]
@chriseth updated this myself

**Review by chriseth — APPROVED** [review]

**Pull request merged**